### DISCUSSION TITLE
PEEK & POKE Access to C64/C128 Memory

### DISCUSSION THREAD

**Discussion by RetrofanC** [zeldin/OrangeCart#11] [Q&A] [answered]
Hi, 

Also I am the proud owner of a OrangeCard. 
Really a masterpiece of innovation from Marcus. ((-8 

I am doing some baby steps but have some great idea's. 
But I got already stuck with a c64/c128 memory PEEK. 
Hopefully one of you can help me out. 

I have the following simple test program. 
Changing border colors and a pause loop. 

Now the PEEK only works the 'first time'.
Meaning when I do a; 

POKE 4096,0; 

at the c64/c128 before I do the RVSYS$40000000

I do get the black border as first one and the @ sign at screen location $0500. 
However, when the program is running in an endless while loop on the RISCV side. 
I would have expected that when I do another; 

POKE 4096,1; 

at the c64/c128 side, 
it will be picked up in one of the "while" cycles and changes the color to white and character at $0500 to A. 

But that doesn't happen....... )-8 

Only when I do; 

POKE 4096,1; 
RVSYS$40000000

The program starts again and shows the white border and A character as first one.  

What is going wrong? 
- Am I making a thinking mistake about the architecture? 
- Do I need to add something to the c-code? 

I also tried to use; 
#define PEEK(addr) (*((const volatile uint8_t *)(void *)(addr)))
which doesn't make a difference. 

I am sure it is something simple but I have tried a lot already and are just overlooking it. 
So I thought lets check the community!
They might be able to point me in the right direction.... please.. 

=============================================================
#include <stdint.h>

#define PEEK(addr) (*((volatile uint8_t *)(void *)(addr)))
#define POKE(addr,value) do{*((volatile uint8_t *)(void *)(addr))=(value);}while(0)

void main(void)
{
 volatile uint32_t count; 
 volatile uint8_t color; 
 color = 2;

 while(1){ 
   color = PEEK(0x1000);

   for(count=0; count < 10000000;count++); 
   POKE(0xD020,color);
   POKE(0x0500,color);

   for(count=0; count < 10000000;count++); 
   POKE(0xD020,0x03);
   POKE(0x0500,0x03);

   for(count=0; count < 10000000;count++); 
   POKE(0xD020,0x04);
   POKE(0x0500,0x04);

   for(count=0; count < 10000000;count++); 
   POKE(0xD020,0x05);
   POKE(0x0500,0x05);

   for(count=0; count < 10000000;count++); 
   POKE(0xD020,0x06);
   POKE(0x0500,0x06);

   for(count=0; count < 10000000;count++); 
   POKE(0xD020,0x07);
   POKE(0x0500,0x07);

   for(count=0; count < 10000000;count++); 
   POKE(0xD020,0x08);
   POKE(0x0500,0x08);

   for(count=0; count < 10000000;count++); 
 }
 for(;;);
}

**Comment by zeldin** [comment, marked as answer]
Because the C64 memory space is in the address range 0x0000-0xFFFF, it falls in the range covered by the RiscV core's data cache (any address with the MSB not set is cacheable in the simple setup used).  Thus, once you read a value it will end up in the data cache, and stay there until that cache line gets evicted.  The `RVSYS` command flushes both the data and instruction cache before starting the code, which is why re-issuing the `RVSYS` command fixes the problem for you.

You can use this macro to flush the data cache from inside your RiscV code:
```
#define CACHE_FLUSH() __asm__ __volatile__(".word 0x500F")
```

However, if you only want to pass some small message from the C64 side to the RiscV, a better option may be to use the mailbox area, which is uncached on the RiscV side.  Please see [the documentation](https://github.com/zeldin/RVCop64/blob/master/doc/ioregs.md#mailbox) for details.

**Reply by RetrofanC** [reply]
Brilliant, works like a charm now, thanks!
